[SE-0198] Introduced CustomPlaygroundDisplayConvertible and deprecated PlaygroundQuickLook/CustomPlaygroundQuickLookable

cwakamo opened the pull request:
Introduced the `CustomPlaygroundDisplayConvertible` protocol and unconditionally deprecated `PlaygroundQuickLook` and `CustomPlaygroundQuickLookable` (and all conformances to those protocols) as part of implementation of SE-0198.

These are the Swift-side changes for &lt;rdar://problem/32905293&gt;.

cwakamo committed:
[stdlib] Introduced the `CustomPlaygroundDisplayConvertible` protocol.

This protocol is intended as a replacement for `CustomPlaygroundQuickLookable`
and `PlaygroundQuickLook`. It allows a type to return a substitute instance for
use during playground logging.

cwakamo committed:
[stdlib] Deprecated `PlaygroundQuickLook` and `CustomPlaygroundQuickLookable`.

Deprecated the `PlaygroundQuickLook` enum and `CustomPlaygroundQuickLookable`
protocol. These are being targeted for removal in Swift 5, so we want to
unconditionally deprecate them now to encourage use of
`CustomPlaygroundDisplayConvertible` instead.

This commit includes deprecated the various `CustomPlaygroundQuickLookable`
conformances across the standard library and overlay libraries.

cwakamo commented:
Getting this up now so I can do this:

@swift-ci Please test

cwakamo commented:
@airspeedswift Would you mind reviewing this to make sure I'm not off-base with how I put this in and the wording of the deprecation warning? (Or, alternatively, suggest someone else who can review it instead?)

moiseev commented on stdlib/public/SDK/AppKit/AppKit.swift:
```diff
@@ -14,6 +14,7 @@ import Foundation
 @_exported import AppKit
 
 extension NSCursor : _DefaultCustomPlaygroundQuickLookable {
+  @available(*, deprecated, message: "NSCursor._defaultCustomPlaygroundQuickLook will be removed in a future Swift version")
```
I think we should still allow using it in Swift 3 and Swift 4.1 without warning. Therefore the attribute should look like:
`@available(swift, deprecated: 4.1, message: "...")`.

natecook1000 commented on stdlib/public/core/Mirror.swift:
```diff
@@ -660,8 +660,22 @@ internal extension Mirror {
 
 //===--- QuickLooks -------------------------------------------------------===//
 
-/// The sum of types that can be used as a Quick Look representation.
```
Let's keep the old abstract here and in the next one—the `@availability` deprecation gives enough information to mark the symbol.

natecook1000 commented on stdlib/public/core/Mirror.swift:
````diff
@@ -660,8 +660,22 @@ internal extension Mirror {
+/// #endif
+/// ```
+///
+/// - seealso: `CustomPlaygroundDisplayConvertible`
````
Can you drop these `seealso` markers, too?

natecook1000 commented on stdlib/public/core/Mirror.swift:
````diff
@@ -660,8 +660,22 @@ internal extension Mirror {
+/// present when compiling with Swift 4.0 or Swift 3.2 or earlier:
+///
+/// ```swift
+/// #if !(swift(>=4.1) || swift(>=3.3) && !swift(>=4.0))
````
Our Markdown style for examples is to use indented code blocks, not fenced ones, unless we're writing non-Swift code. The code itself should be indented with 4 spaces.

natecook1000 commented on stdlib/public/core/PlaygroundDisplay.swift:
```diff
@@ -0,0 +1,63 @@
+    ///
+    /// If this type has value semantics, the instance returned should be
+    /// unaffected by subsequent mutations if possible.
+    var playgroundDescription: Any { get }
```
This block should only be indented two spaces.

cwakamo committed:
[stdlib] Updated the doc comments for `PlaygroundQuickLook`/`CustomPlaygroundQuickLookable`.

Since these are now deprecated (in preparation for removal in a future Swift
release), the doc comments reflect that and provide suggestions for what to do
instead (i.e. use `CustomPlaygroundDisplayConvertible`) and how to
conditionalize usage of `PlaygroundQuickLook`/`CustomPlaygroundQuickLookable` so
that source code remains compilable both by old versions which only have the old
protocol and newer versions which have removed the old protocol.

cwakamo commented:
@swift-ci Please smoke test

cwakamo commented:
@natecook1000 I added back the old doc comments but left the deprecation notices as notes so that I could show how to properly conditionalize usage of `CustomPlaygroundQuickLookable`/`PlaygroundQuickLook` so that code continues to compile when we're able to remove them from the standard library. If you think that's overkill, I can drop my additions to those comments entirely.

natecook1000 reviewed (approved):
@cwakamo LGTM, thanks!
